0001-data-collection-extension/README.md

```diff
@@ -6,62 +6,56 @@ editor: Alexey Shmalko <ashmalko@cybervisiontech.com>
-
-2. A server should handle different types of data. At one time point the endpoint can send one type of data and at another time point it can send different one. The variety of data types are represented via multiple data formats (JSON, Protocol Buffers) and different data itself.
-
-  - The issue can be resolved using separate address tokens for all the combinations of data formats with data set. The mean for data sets descriptions is data schemes. A data schemes set and a data formats set should be restricted by an [application version]() e.g. if we have the application v.1.0.0, it supports JSON data format and data schema v.1 then the application can not send data in protobuf format or use another data schema version. For this purpose another application should be created, with another application version. Possible format for address token is:
-
-    ```
-    <endpoint>/<format>[/<schema>]
-    ```
+1. _Record processing is asynchronous._ Processing might require a significant period of time and may be performed simultaneously or in different order.
 
-3. Sometimes data from different endpoint should be handled the different way and a server has no notion of which endpoint sends data to it.
+2. _The client should know if a record has been delivered and/or processed._ This is needed for client to guarantee data has been processed and free resources.
 
-  - The solution is simple: we can use `<endpoint>` part in the address token to uniquely identify an endpoint.
+   Solutions:
+   - Use QoS levels for message delivery confirmation.
+   - Use QoS levels for message processing confirmation.
 
-4. Internet connection can be slow or unstable or really expensive to waste the traffic.
+     MQTT requires all PUBACK messages appear in the same order as received PUBLISH messages. This creates a synchronization point, which conflicts with asynchronous record processing.
+   - Use separate response messages for processing confirmation
 
-  - There is a need to send as little data as possible. Because of the fact that a logs amount can be sufficient, even in the case with stable connection, adding some excessive meta-data increases overhead in a significant degree. So, possible solutions are:
+     For CoAP, this is supported by [Separate Response (RFC 7252, Section 5.2.2)](https://tools.ietf.org/html/rfc7252#section-5.2.2).
 
-    - Using *batches*. In this case meta-data is added to batches instead of log records and log records are uploaded batch-by-batch, so we save a lot of internet traffic.
-    - Using *delta compression*.
+     For MQTT, this is achieved by publishing a response into the status topic. (See [Kaa Protocol](/0002-kaa-protocol/README.md).)
 
-5. Devices can have no ability to generate timestamps.
+3. _A server should handle different types of data._ Different endpoints may send data in a variety of formats; the server should know the format to parse the payload.
```

Comment:
Basically, that's not even true. The available formats are restricted by application version and extension configuration, not endpoint token. (Basically, it might be restricted by endpoint token, but that's well out of scope for now.)